tensorflow/contrib/opt/python/training/elastic_average_optimizer.py

```diff
@@ -246,6 +296,25 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
       local_update = state_ops.assign_add(
           self._local_step, 1, name='local_step_update').op
 
+    # this is for place the variables created by optimizer to local collection
+    # e.g., AdamOptimizer will create beta as global variables
+    def _adjust_optimizer_variable_collection():
```

Comment:
Why is this necessary?

Reply:
When do the apply_gradients, optimizers like MomentumOptimizer/AdamOptimizer/RMSPropOptimizer will created new variables and put to GLOBAL_VARIABLES collection. While, these variables in fact belongs to local models, local worker will update local model and these newly created variables.

Reply:
I'm not sure I follow. I think this change is trying to do too much, and this will be fragile unless this optimizer is used in precisely the right way.

I'd like you to find less intrusive approaches to achieve this, or more invasive approaches that prescribe the code to be written in a certain way for elastic averaging to work.

Reply:
As it's for the variables created by apply_gradients only, I'm considering only take care of the incremental part, instead of do it on the whole GLOBAL_VARIABLES collection. How does this sound to you?

Reply:
Yes, this is better.